nuxeo-runtime/nuxeo-runtime-test/src/main/java/org/nuxeo/runtime/test/RuntimeHarnessImpl.java

```diff
@@ -0,0 +1,478 @@
+import org.nuxeo.runtime.osgi.OSGiRuntimeService;
+import org.nuxeo.runtime.test.protocols.inline.InlineURLFactory;
+import org.nuxeo.runtime.test.runner.RuntimeHarness;
+import org.nuxeo.runtime.test.runner.TargetExtensions;
+import org.nuxeo.runtime.transaction.TransactionHelper;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkEvent;
+
+import io.github.lukehutch.fastclasspathscanner.FastClasspathScanner;
+
+public final class RuntimeHarnessImpl implements RuntimeHarness {
+
+    protected static final Log log = LogFactory.getLog(RuntimeHarnessImpl.class);
+
+    protected static URL[] introspectClasspath(ClassLoader loader) {
+        return new FastClasspathScanner().getUniqueClasspathElements().stream().map(file -> {
+            try {
+                return file.toURI().toURL();
+            } catch (MalformedURLException cause) {
+                throw new Error("Could not get URL from " + file, cause);
+            }
+        }).toArray(URL[]::new);
+    }
+
+    protected StandaloneBundleLoader bundleLoader;
+
+    protected Map<String, BundleFile> bundles;
+
+    protected boolean frameworkStarted;
+
+    protected OSGiAdapter osgi;
+
+    protected Set<URI> readUris;
+
+    protected OSGiRuntimeService runtime;
+
+    protected Bundle runtimeBundle;
+
+    protected final TargetResourceLocator targetResourceLocator;
+
+    protected URL[] urls; // classpath urls, used for bundles lookup
+
+    protected final List<WorkingDirectoryConfigurator> wdConfigs;
+
+    protected File workingDir;
+
+    public RuntimeHarnessImpl(Class<?> clazz) {
+        targetResourceLocator = new TargetResourceLocator(clazz);
+        wdConfigs = new ArrayList<>();
+    }
+
+    @Override
+    public void addWorkingDirectoryConfigurator(WorkingDirectoryConfigurator config) {
+        wdConfigs.add(config);
+    }
+
+    @Override
+    public void deployBundle(String name) throws Exception {
+        // install only if not yet installed
+        Bundle bundle = bundleLoader.getOSGi().getRegistry().getBundle(name);
+        if (bundle == null) {
+            BundleFile bundleFile = lookupBundle(name);
+            bundleLoader.loadBundle(bundleFile);
+            bundleLoader.installBundle(bundleFile);
+            bundle = bundleLoader.getOSGi().getRegistry().getBundle(name);
+        }
+        if (runtime.getContext(bundle) == null) {
+            runtime.createContext(bundle);
+        }
+    }
+
+    @Override
+    public void deployContrib(String name, String contrib) throws Exception {
+        RuntimeContext context = runtime.getContext(name);
+        if (context == null) {
+            context = runtime.getContext();
+            BundleFile file = lookupBundle(name);
+            URL location = file.getEntry(contrib);
+            if (location == null) {
+                throw new AssertionError("Cannot locate " + contrib + " in " + name);
+            }
+            context.deploy(location);
+            return;
+        }
```

Comment:
Could you use if/else instead of if/return ?

Reply:
Definitely, code has been c/c, done.